language/stdlib/modules/Libra.move

```diff
@@ -700,71 +700,159 @@ module Libra {
+        invariant module forall coinType: type, addr: address:
+            !spec_is_currency<coinType>() ==> !exists<BurnCapability<coinType>>(addr);
+
+        /// After a currency is registered, only the treasury compliance account
+        /// has the mint capability for that currency.
+        invariant module forall coinType: type, addr: address
+        where spec_is_currency<coinType>():
+            exists<BurnCapability<coinType>>(addr)
+                ==> addr == spec_treasury_compliance_addr();
+    }
+
+    spec module {
+        /// Similar to OnlyTCHasMintCapabilityInvariant, this invariant
+        /// cannot be proved right now.
+        apply OnlyTCHasBurnCapabilityInvariant to *, *<CoinType>;
     }
 
-    // Maintain a ghost variable representing the sum of
-    // all coins of a currency type.
+    /// ## Conservation of currency
+
     spec module {
+        /// Maintain a ghost variable representing the sum of
+        /// all coins of a currency type.
         global sum_of_coin_values<CoinType>: num;
     }
+
     spec struct Libra {
         invariant pack sum_of_coin_values<CoinType> = sum_of_coin_values<CoinType> + value;
         invariant unpack sum_of_coin_values<CoinType> = sum_of_coin_values<CoinType> - value;
     }
 
+    spec schema TotalValueRemainsSame<CoinType> {
+        /// The total amount of currency stays constant.
+        ensures sum_of_coin_values<CoinType> == old(sum_of_coin_values<CoinType>);
+    }
+
+    spec module {
+       /// Only mint/burn & with capability versions can change the total amount of currency.
+       apply TotalValueRemainsSame<CoinType> to *<CoinType> except mint*<CoinType>, burn*<CoinType>;
+    }
+
+    spec schema SumOfCoinValuesInvariant<CoinType> {
+        /// The sum of value of coins is consistent with
+        /// the total_value CurrencyInfo keeps track of.
+        invariant module !spec_is_currency<CoinType>() ==> sum_of_coin_values<CoinType> == 0;
+        invariant module spec_is_currency<CoinType>()
+                    ==> sum_of_coin_values<CoinType>
+                        == global<CurrencyInfo<CoinType>>(spec_currency_addr()).total_value;
+    }
+
+   spec module {
+       apply SumOfCoinValuesInvariant<CoinType> to public *<CoinType> except is_currency<CoinType>;
+   }
+
+    /**
+    TODO: specify the following:
+
+          sum of coin values in preburn + sum of account balances == total value
+
+          This is false right now because in `cancel_burn`, we return the coin
+          directly instead of depositing it into an account. However, even after
+          this has been fixed, we will likely need to coordinate with the account
+          module and maybe add a ghost variable representing the sum of account
+          balances.
+    **/
+
     // TODO: What happens to the CurrencyRegistrationCapability?
 
 
+    /// ## Specifications for individual functions
+
+    spec schema AbortsIfSenderIsNotRoot {
```

Comment:
Don't worry to much about local specs. I will take care of this, porting the schema-based older Libra spec which we need to decommission anyway.